### ISSUE TITLE
Create something akin to `vimtutor` for Helix

### ISSUE THREAD

**Issue by Omnikar** [helix-editor/helix#506]
This would introduce the user to modal editing, and to Helix's commands and to its multiple selections. I think it could be opened via something like `hx --tutor`, or opened from within the editor with something like `:tutor`.

**Comment by python128** [comment]
VIm tutor is basically a text file which contains explanations of keys, and some examples to work on, right?
If so, can I have a go at this?

**Comment by Omnikar** [comment]
Yes, `vimtutor` is just a text file separated into lessons which interactively introduce the keys and concepts, and each of which is sized to fit on an 80x24 terminal screen.

**Comment by aeosynth** [comment]
Kakoune has something similar: https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE

**Comment by Omnikar** [comment]
> Kakoune has something similar: https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE

I like that file header, and sorta wanna steal that idea…
```
 _   _                 __    __
| | | |         _      \ \  / /
| |_| |   ___  | | (_)  \ \/ /
|  _  |  / _ \ | |  _    )  (
| | | | |  __/ | | | |  / /\ \
|_| |_|  \___| |_| |_| /_/  \_\
```

**Comment by Omnikar** [comment]
> Kakoune has something similar: https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE

I know this doesn't have to do with the tutorial itself, but what if we were to copy that title page idea?
```
 _   _                 __    __
| | | |         _      \ \  / /
| |_| |   ___  | | (_)  \ \/ /
|  _  |  / _ \ | |  _    )  (
| | | | |  __/ | | | |  / /\ \
|_| |_|  \___| |_| |_| /_/  \_\
```

**Comment by python128** [comment]
> > Kakoune has something similar: https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE
> 
> I know this doesn't have to do with the tutorial itself, but what if we were to copy that title page idea?
> 
> ```
>  _   _                 __    __
> | | | |         _      \ \  / /
> | |_| |   ___  | | (_)  \ \/ /
> |  _  |  / _ \ | |  _    )  (
> | | | | |  __/ | | | |  / /\ \
> |_| |_|  \___| |_| |_| /_/  \_\
> ```

Actually a good idea, but I would suggest using some other format, rather than the exact one used by Kakoune...

**Comment by Omnikar** [comment]
> Kakoune has something similar: https://github.com/mawww/kakoune/blob/master/contrib/TRAMPOLINE

I don't think I like the format of this in comparison to Vimtutor. The latter emphasizes that the user should learn by doing, rather than trying to memorize everything, and provides tasks for the user to complete using what is taught in each lesson. Whereas Kakoune's seems to just list the keys and describe what they do.

**Comment by python128** [comment]
How about I do it like `vimtutor`, but some ideas used by Kakoune's list. Eg:
I can make it interactive like `vimtutor`, but make the keys like kakoune. i.e. :

```
       .---,
       | ↑ |        Movement in a buffer (the representation of the contents
   .---'---'---,    of a file opened by Kakoune) can be achieved using the arrow
   | ← | ↓ | → |    keys, which will move the cursor up one column/row into
   `---'---'---`    a given direction.
```

**Comment by Omnikar** [comment]
As long as it maintains Vimtutor's premise of learning by doing by giving the user small tasks to perform using the skills they learn. Also, you might want to open a draft PR so that you can receive feedback as you work on this.

**Comment by olanod** [comment]
As someone who finds the project quite neat but a bit clueless about its full potential and advance usage I'll just stick around to watch how tutorials and documentation get better 👀

**Comment by Omnikar** [comment]
Ok, a partial tutorial has been added by #537. I'll leave this open until the tutorial is completed.

**Comment by A-Walrus** [comment]
Should there be some kind of explanation on using LSP and TS features within tutor? 

TS features should be fairly straightforward, show some example code in whatever language, and ask them to navigate with go to next function and things like that.

For the language server it's a bit tricky since we can't really rely on the user having any specific language server installed can we? 
I would want an explanation of capabilities such as go to definition/implementation/etc..., show documentation, code actions.

**Comment by the-mikedavis** [comment]
Both are difficult to add to the current tutor because they need the document to be a language with tree-sitter and language-server support and currently it's text, plus it would need you to install the language server for that language. That may be better left off as a separate git project similar to https://github.com/rust-lang/rustlings

**Comment by CobaltCause** [comment]
It'd be cool if the tutor file was somehow templated from the configured keybinds so it would always display the correct instructions even if the user had rebound the key. I imagine this would be pretty difficult to implement though.

**Comment by correabuscar** [comment]
in `3.5 SELECT / EXTEND MODE`, if I mouse click on `B` of `FOO BAR` then press `b` to get to `F`, because `FOO ` is now selected, then when I do the suggested `v2w` it selects only ` BAR `. Maybe mentioning the solution for to how to "fix" this possibility would be great to see there. 

I might be the only one, but this kind of thing stops me in my tracks and I stop following the tutor and do something else, unrelated, instead. ie. it's easy to give up on helix when tiny things like this happen, and just fall back to nvim.

**Comment by the-mikedavis** [comment]
That's https://github.com/helix-editor/helix/issues/3842.

**Comment by divarvel** [comment]
Does the tutor file have a specific language set? The match mode works only partially in it

- `mm` does not work
- `mam` / `mim` work with brackets, braces, parens, but not double quotes
- `ms` works
- `md` / `mr` don't work

I’m a bit surprised that `mam`/`mim` work for some pairs, but not quotes, and that `mm` does not work at all, even for pairs supported by `mam` / `mim`.

That makes writing a chapter on the match minor mode a bit tricky.

**Comment by goyalyashpal** [comment]
> if the tutor file was somehow templated from the configured keybinds
\- @ CobaltCause at https://github.com/helix-editor/helix/issues/506#issuecomment-1233782124

i was thinking exact same thing :)

**Comment by darknebula05** [comment]
On line 983 of the tutor it says to press <C-s> to save a position to the jumplist. However, when I looked at the keybinding using the command palett it says <C-s> is bound to `:write`

**Comment by jokeyrhyme** [comment]
Perhaps the TS and LSP lessons could be implemented in multiple languages, and `:tutor` could either prompt the user to pick, or could auto-select one based on the best available combination of lesson + grammar + server?

There's probably only a handful of languages needed to significantly cover the user base

**Comment by jokeyrhyme** [comment]
We could also wait for the plugin system, and either ship the TS and LSP for the plugin system's language, or privilege that language by having `:tutor` show you how to install any missing pieces?

**Comment by PaneradFisk** [comment]
I'm doing the tutor. I know I'm nitpicking... sorry! D:
It seems written biased to _US Qwerty keyboard layout_.
On line 498 it says  `Note: To search backwards, type ? (Shift-/).`
I'm on a _Nordic Qwerty International keyboard layout_. It looks and works completely different.
(? on my keyboard is Shift and +, where + is next to 0)

Might want to remove such hints, as they would probably cause more confusion than anything for someone not well versed in different keyboard layouts.
Thank you!

See picture for comparison
![image](https://github.com/user-attachments/assets/f4b1e4a4-7804-45fa-8900-62d7a6a01a58)
